akka-cluster-typed/src/main/scala/akka/cluster/ddata/typed/javadsl/DistributedData.scala

```diff
@@ -37,10 +40,34 @@ object DistributedData extends ExtensionId[DistributedData] {
 abstract class DistributedData extends Extension {
 
   /**
-   * `ActorRef` of the [[Replicator]] .
+   * `ActorRef` of the [[Replicator]].
+   *
+   * @see [[DistributedData.replicatorMessageAdapter]]
    */
   def replicator: ActorRef[Replicator.Command]
 
+  /**
+   * When interacting with the [[DistributedData.replicator]] from an actor the [[ReplicatorMessageAdapter]]
+   * provides convenient methods that adapts the response messages to the requesting actor's message protocol.
+   *
+   * One `ReplicatorMessageAdapter` instance can be used for a given `ReplicatedData` type,
+   * e.g. an `OrSet<String>`. Interaction with several [[akka.cluster.ddata.Key]]s can be used via the same adapter
+   * but they must all be of the same `ReplicatedData` type. For interaction with several different
+   * `ReplicatedData` types, e.g. an `OrSet<String>` and a `GCounter`, an adapter can be created
+   * for each type.
+   *
+   * *Warning*: `ReplicatorMessageAdapter` is not thread-safe and must only be used from the actor
+   * corresponding to the given `ActorContext`. It must not be accessed from threads other
+   * than the ordinary actor message processing thread, such as [[java.util.concurrent.CompletionStage]]
+   * callbacks. It must not be shared between several actor instances.
+   *
+   * @param context The [[ActorContext]] of the requesting actor. The `ReplicatorMessageAdapter` can only be
+   *                used in this actor.
+   * @tparam A Message type of the requesting actor.
+   * @tparam B Type of the [[ReplicatedData]].
+   */
+  def replicatorMessageAdapter[A, B <: ReplicatedData](context: ActorContext[A]): ReplicatorMessageAdapter[A, B]
```

Comment:
Maybe too late for coming in and dropping ideas but, if it was a behavior decorator rather than this style you'd be hard pressed to do it in an unsafe way, and also it would align with other resourcey:things (mdc, timers etc)

Pseudo-code-idea (names are crap):
```
ReplicatorAdapter(key){ adapter => 
  adapter.ask etc
  Behavior.stuff
}
```

Reply:
@johanandren I like that idea. I'll give it a try but let's not halt the M3 release because of this PR. It can be included in M4 instead.

Reply:
Sounds good to me!

Reply:
@johanandren Changed to `with` approach in deaf45c09f889c567d75e892ddc8f75bedb6530c